core/src/main/java/org/apache/calcite/rel/rules/SemiJoinRule.java

```diff
@@ -232,6 +234,87 @@ default JoinToSemiJoinRuleConfig withOperandFor(Class<Join> joinClass,
     }
   }
 
+  /**
+   * SemiJoinRule that matches a Project on top of a Join with a RelNode
+   * which is unique for Join's right keys.
+   *
+   * @see CoreRules#JOIN_ON_UNIQUE_TO_SEMI_JOIN */
+  public static class JoinOnUniqueToSemiJoinRule extends SemiJoinRule {
+
+    /** Creates a JoinOnUniqueToSemiJoinRule. */
+    protected JoinOnUniqueToSemiJoinRule(JoinOnUniqueToSemiJoinRuleConfig config) {
+      super(config);
+    }
+
+    @Override public boolean matches(RelOptRuleCall call) {
+      final Project project = call.rel(0);
+      final Join join = call.rel(1);
+      final RelNode left = call.rel(2);
+
+      final ImmutableBitSet bits =
+          RelOptUtil.InputFinder.bits(project.getProjects(), null);
+      final ImmutableBitSet rightBits =
+          ImmutableBitSet.range(left.getRowType().getFieldCount(),
+              join.getRowType().getFieldCount());
+      return !bits.intersects(rightBits);
+    }
+
+    @Override public void onMatch(RelOptRuleCall call) {
+      final Project project = call.rel(0);
+      final Join join = call.rel(1);
+      final RelNode left = call.rel(2);
+      final RelNode right = call.rel(3);
+
+      // Same with ProjectToSemiJoinRule
+      if (right instanceof Aggregate) {
```

Comment:
Should we rather add in the new rule's config a predicate to not match Aggregates on the join RHS:
```
...
b2 -> b2.operand(Join.class).predicate(SemiJoinRule::isJoinTypeSupported).inputs(
  b3 -> b3.operand(RelNode.class).anyInputs(),
  b4 -> b4.operand(RelNode.class).predicate(n -> !(n instanceof Aggregate)).anyInputs()
)))
```
Rather than matching it with two rules (the old and the new) and creating two times the same result?

Reply:
Good idea!

Reply:
Thanks for the reactivity @libenchao !
Nitpick: perhaps adding a comment in the predicate, to clarify that the new rule does not match aggregates on the join RHS because this case is already covered by ProjectToSemiJoinRule ?

Reply:
sounds good, added.

Reply:
Thanks